File: src/test/java/org/hisp/dhis/rules/RuleEngineContextTest.java
package org.hisp.dhis.rules;

import org.hisp.dhis.rules.models.Rule;
import org.hisp.dhis.rules.models.RuleVariable;
import org.hisp.dhis.rules.util.MockRuleVariable;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.fail;

@RunWith( JUnit4.class )
public class RuleEngineContextTest
{
    private final RuleVariable ruleVariable = new MockRuleVariable();

    private final RuleVariable ruleVariableTwo = new MockRuleVariable();

    private final Rule rule = Rule.MOCK;

    private final Rule ruleTwo = Rule.MOCK;
    
    @Test( expected = IllegalArgumentException.class )
    public void builderShouldThrowOnNullVariableList()
    {
        RuleEngineContext.builder()
            .rules( new ArrayList<Rule>() )
            .ruleVariables( null );
    }

    @Test( expected = IllegalArgumentException.class )
    public void builderShouldThrowOnNullRulesList()
    {
        RuleEngineContext.builder()
            .ruleVariables( new ArrayList<RuleVariable>() )
            .ruleVariables( null );
    }

    @Test
    public void builderShouldContainImmutableCopyOfRules()
    {
        List<String> members = List.of( "one", "two" );
        Map<String, List<String>> supplementaryData = new HashMap<>();
        supplementaryData.put( "text-key", members );

        List<RuleVariable> ruleVariables = new ArrayList<>();
        List<Rule> rules = new ArrayList<>();

        ruleVariables.add( ruleVariable );
        rules.add( rule );

        RuleEngineContext ruleEngineContext = RuleEngineContext.builder()
            .ruleVariables( ruleVariables )
            .supplementaryData( supplementaryData )
            .rules( rules )
            .build();

        ruleVariables.add( ruleVariableTwo );
        rules.add( ruleTwo );

        assertEquals( 1 , ruleEngineContext.ruleVariables().size() );
        assertEquals( ruleVariable , ruleEngineContext.ruleVariables().get( 0 ) );

        assertEquals( 1 , ruleEngineContext.supplementaryData().size() );
        assertNotNull( ruleEngineContext.supplementaryData().get( "text-key" ) );
        assertEquals( members , ruleEngineContext.supplementaryData().get( "text-key" ) );

        assertEquals( 1 , ruleEngineContext.rules().size() );
        assertEquals( rule , ruleEngineContext.rules().get( 0 ) );

        try
        {
            ruleEngineContext.ruleVariables().clear();
            fail( "UnsupportedOperationException was expected, but nothing was thrown." );
        }
        catch ( UnsupportedOperationException unsupportedOperationException )
        {
            // noop
        }

        try
        {
            ruleEngineContext.rules().clear();
            fail( "UnsupportedOperationException was expected, but nothing was thrown." );
        }
        catch ( UnsupportedOperationException unsupportedOperationException )
        {
            // noop
        }
    }

    @Test
    public void toEngineBuilderShouldReturnNewInstances()
    {
        RuleEngineContext ruleEngineContext = RuleEngineContext.builder()
            .ruleVariables(List.of(new MockRuleVariable()))
            .supplementaryData( new HashMap<String, List<String>>() )
            .rules( List.of(Rule.MOCK) )
            .build();

        RuleEngine.Builder ruleEngineBuilderOne = ruleEngineContext.toEngineBuilder();
        RuleEngine.Builder ruleEngineBuilderTwo = ruleEngineContext.toEngineBuilder();

        assertNotEquals( ruleEngineBuilderOne, ruleEngineBuilderTwo );
    }
}

File: src/test/java/org/hisp/dhis/rules/RuleEngineContextTest.kt
package org.hisp.dhis.rules

import io.mockk.mockk
import org.hisp.dhis.rules.models.Rule
import org.hisp.dhis.rules.models.RuleVariable
import org.hisp.dhis.rules.util.MockRuleVariable
import org.junit.Assert
import org.junit.Test

class RuleEngineContextTest {
    private val ruleVariable: RuleVariable = MockRuleVariable()
    private val ruleVariableTwo: RuleVariable = MockRuleVariable()
    private val rule: Rule = mockk<Rule>()
    private val ruleTwo: Rule = mockk<Rule>()
    @Test(expected = IllegalArgumentException::class)
    fun builderShouldThrowOnNullVariableList() {
        RuleEngineContext.builder()
            .rules(ArrayList())
            .ruleVariables(null)
    }

    @Test(expected = IllegalArgumentException::class)
    fun builderShouldThrowOnNullRulesList() {
        RuleEngineContext.builder()
            .ruleVariables(ArrayList())
            .ruleVariables(null)
    }

    @Test
    fun builderShouldContainImmutableCopyOfRules() {
        val members = listOf("one", "two")
        val supplementaryData: MutableMap<String, List<String>> = HashMap()
        supplementaryData["text-key"] = members
        val ruleVariables: MutableList<RuleVariable> = ArrayList<RuleVariable>()
        val rules: MutableList<Rule> = ArrayList()
        ruleVariables.add(ruleVariable)
        rules.add(rule)
        val ruleEngineContext = RuleEngineContext.builder()
            .ruleVariables(ruleVariables)
            .supplementaryData(supplementaryData)
            .rules(rules)
            .build()
        ruleVariables.add(ruleVariableTwo)
        rules.add(ruleTwo)
        Assert.assertEquals(1, ruleEngineContext.ruleVariables.size.toLong())
        Assert.assertEquals(ruleVariable, ruleEngineContext.ruleVariables[0])
        Assert.assertEquals(1, ruleEngineContext.supplementaryData.size.toLong())
        Assert.assertNotNull(ruleEngineContext.supplementaryData["text-key"])
        Assert.assertEquals(members, ruleEngineContext.supplementaryData["text-key"])
        Assert.assertEquals(1, ruleEngineContext.rules.size.toLong())
        Assert.assertEquals(rule, ruleEngineContext.rules[0])
        try {
            ruleEngineContext.ruleVariables.clear()
            Assert.fail("UnsupportedOperationException was expected, but nothing was thrown.")
        } catch (unsupportedOperationException: UnsupportedOperationException) {
            // noop
        }
        try {
            ruleEngineContext.rules.clear()
            Assert.fail("UnsupportedOperationException was expected, but nothing was thrown.")
        } catch (unsupportedOperationException: UnsupportedOperationException) {
            // noop
        }
    }

    @Test
    fun toEngineBuilderShouldReturnNewInstances() {
        val ruleEngineContext = RuleEngineContext.builder()
            .ruleVariables(listOf<RuleVariable>(MockRuleVariable()))
            .supplementaryData(HashMap<String, List<String>>())
            .rules(listOf(mockk<Rule>()))
            .build()
        val ruleEngineBuilderOne = ruleEngineContext.toEngineBuilder()
        val ruleEngineBuilderTwo = ruleEngineContext.toEngineBuilder()
        Assert.assertNotEquals(ruleEngineBuilderOne, ruleEngineBuilderTwo)
    }
}
